File: src/web-ui/src/infrastructure/api/service-api/SpeechAPI.ts
import { api } from './ApiClient';
import { createTauriCommandError } from '../errors/TauriCommandError';

export const LOCAL_SENSEVOICE_SMALL_INT8_MODEL_ID = 'sensevoice-small-int8';
export const LOCAL_QWEN3_ASR_0_6B_INT8_MODEL_ID = 'qwen3-asr-0.6b-int8';
export const DEFAULT_SPEECH_SAMPLE_RATE = 16000;
export const DEFAULT_MAX_RECORDING_SECONDS = 60;
export const SPEECH_MODEL_PROGRESS_EVENT = 'speech://model-download-progress';
export const SPEECH_MODEL_STATUS_CHANGED_EVENT = 'speech://model-status-changed';

export type SpeechModelInstallState =
  | 'not_installed' | 'downloading' | 'installed' | 'verifying'
  | 'corrupt' | 'deleting' | 'error';

export interface SpeechModelProgress {
  modelId: string;
  downloadedBytes: number;
  totalBytes: number;
  percent: number;
}

export interface SpeechModelStatus {
  modelId: string;
  displayName: string;
  provider: string;
  version: string;
  description: string;
  languages: string[];
  state: SpeechModelInstallState;
  installedPath?: string | null;
  installedBytes: number;
  expectedBytes: number;
  progress?: SpeechModelProgress | null;
  error?: string | null;
}

export interface SpeechListModelsResponse { models: SpeechModelStatus[]; }
export interface SpeechModelProgressEvent { status: SpeechModelStatus; }
export interface SpeechStartInputSessionRequest {
  modelId?: string | null;
  language?: string | null;
  sampleRate?: number | null;
  maxRecordingSeconds?: number | null;
}
export interface SpeechInputSession {
  sessionId: string;
  modelId: string;
  language: string;
  sampleRate: number;
  maxRecordingSeconds: number;
}
export interface SpeechAppendAudioChunkResponse {
  receivedBytes: number;
  receivedSeconds: number;
  limitReached: boolean;
}
export interface SpeechTranscriptionResult {
  text: string;
  language: string;
  durationMs: number;
  audioDurationSeconds: number;
}

export class SpeechAPI {
  async listModels() {
    try { return await api.invoke<SpeechListModelsResponse>('speech_list_models', {}); }
    catch (error) { throw createTauriCommandError('speech_list_models', error); }
  }

  async downloadModel(modelId: string) {
    try { return await api.invoke<SpeechModelStatus>('speech_download_model', { request: { modelId } }, { timeout: 600000 }); }
    catch (error) { throw createTauriCommandError('speech_download_model', error, { modelId }); }
  }

  async cancelModelDownload(modelId: string) {
    try { return await api.invoke<SpeechModelStatus>('speech_cancel_model_download', { request: { modelId } }); }
    catch (error) { throw createTauriCommandError('speech_cancel_model_download', error, { modelId }); }
  }

  async deleteModel(modelId: string) {
    try { return await api.invoke<SpeechModelStatus>('speech_delete_model', { request: { modelId } }); }
    catch (error) { throw createTauriCommandError('speech_delete_model', error, { modelId }); }
  }

  async verifyModel(modelId: string) {
    try { return await api.invoke<SpeechModelStatus>('speech_verify_model', { request: { modelId } }); }
    catch (error) { throw createTauriCommandError('speech_verify_model', error, { modelId }); }
  }

  async startInputSession(request: SpeechStartInputSessionRequest = {}) {
    try { return await api.invoke<SpeechInputSession>('speech_start_input_session', { request }); }
    catch (error) { throw createTauriCommandError('speech_start_input_session', error, request); }
  }

  async appendAudioChunk(sessionId: string, pcm16Base64: string) {
    try {
      return await api.invoke<SpeechAppendAudioChunkResponse>('speech_append_audio_chunk', { request: { sessionId, pcm16Base64 } });
    } catch (error) { throw createTauriCommandError('speech_append_audio_chunk', error, { sessionId }); }
  }

  async finishInputSession(sessionId: string) {
    try {
      return await api.invoke<SpeechTranscriptionResult>('speech_finish_input_session', { request: { sessionId } }, { timeout: 120000 });
    } catch (error) { throw createTauriCommandError('speech_finish_input_session', error, { sessionId }); }
  }

  async cancelInputSession(sessionId: string): Promise<void> {
    try { await api.invoke('speech_cancel_input_session', { request: { sessionId } }); }
    catch (error) { throw createTauriCommandError('speech_cancel_input_session', error, { sessionId }); }
  }

  onModelProgress(callback: (event: SpeechModelProgressEvent) => void): () => void {
    return api.listen(SPEECH_MODEL_PROGRESS_EVENT, callback);
  }

  onModelStatusChanged(callback: (status: SpeechModelStatus) => void): () => void {
    return api.listen(SPEECH_MODEL_STATUS_CHANGED_EVENT, callback);
  }
}

export const speechAPI = new SpeechAPI();

File: src/apps/desktop/src/api/speech_api.rs
//! Desktop adapter for local speech input.

use crate::api::AppState;
use bitfun_core_types::speech::{
    SpeechAppendAudioChunkRequest, SpeechAppendAudioChunkResponse, SpeechCancelInputSessionRequest,
    SpeechCancelModelDownloadRequest, SpeechDeleteModelRequest, SpeechDownloadModelRequest,
    SpeechFinishInputSessionRequest, SpeechInputSession, SpeechListModelsResponse,
    SpeechModelProgressEvent, SpeechModelStatus, SpeechStartInputSessionRequest,
    SpeechTranscriptionResult, SpeechVerifyModelRequest,
};
use bitfun_events::{SPEECH_MODEL_PROGRESS_EVENT, SPEECH_MODEL_STATUS_CHANGED_EVENT};
use tauri::{AppHandle, Emitter, State};

#[tauri::command]
pub async fn speech_list_models(
    state: State<'_, AppState>,
) -> Result<SpeechListModelsResponse, String> {
    state
        .speech_service
        .list_models()
        .await
        .map_err(|error| format!("Failed to list speech models: {error}"))
}

#[tauri::command]
pub async fn speech_download_model(
    state: State<'_, AppState>,
    app: AppHandle,
    request: SpeechDownloadModelRequest,
) -> Result<SpeechModelStatus, String> {
    let progress_app = app.clone();
    let status = state
        .speech_service
        .download_model(request, move |event: SpeechModelProgressEvent| {
            if let Err(error) = progress_app.emit(SPEECH_MODEL_PROGRESS_EVENT, &event) {
                log::warn!("Failed to emit speech model progress event: {error}");
            }
        })
        .await
        .map_err(|error| format!("Failed to download speech model: {error}"))?;
    emit_status(&app, &status);
    Ok(status)
}

#[tauri::command]
pub async fn speech_cancel_model_download(
    state: State<'_, AppState>,
    app: AppHandle,
    request: SpeechCancelModelDownloadRequest,
) -> Result<SpeechModelStatus, String> {
    let status = state
        .speech_service
        .cancel_model_download(request)
        .await
        .map_err(|error| format!("Failed to cancel speech model download: {error}"))?;
    emit_status(&app, &status);
    Ok(status)
}

#[tauri::command]
pub async fn speech_delete_model(
    state: State<'_, AppState>,
    app: AppHandle,
    request: SpeechDeleteModelRequest,
) -> Result<SpeechModelStatus, String> {
    let status = state
        .speech_service
        .delete_model(request)
        .await
        .map_err(|error| format!("Failed to delete speech model: {error}"))?;
    emit_status(&app, &status);
    Ok(status)
}

#[tauri::command]
pub async fn speech_verify_model(
    state: State<'_, AppState>,
    app: AppHandle,
    request: SpeechVerifyModelRequest,
) -> Result<SpeechModelStatus, String> {
    let status = state
        .speech_service
        .verify_model(request)
        .await
        .map_err(|error| format!("Failed to verify speech model: {error}"))?;
    emit_status(&app, &status);
    Ok(status)
}

#[tauri::command]
pub async fn speech_start_input_session(
    state: State<'_, AppState>,
    request: SpeechStartInputSessionRequest,
) -> Result<SpeechInputSession, String> {
    state
        .speech_service
        .start_input_session(request)
        .await
        .map_err(|error| format!("Failed to start speech input session: {error}"))
}

#[tauri::command]
pub async fn speech_append_audio_chunk(
    state: State<'_, AppState>,
    request: SpeechAppendAudioChunkRequest,
) -> Result<SpeechAppendAudioChunkResponse, String> {
    state
        .speech_service
        .append_audio_chunk(request)
        .await
        .map_err(|error| format!("Failed to append speech audio chunk: {error}"))
}

#[tauri::command]
pub async fn speech_finish_input_session(
    state: State<'_, AppState>,
    request: SpeechFinishInputSessionRequest,
) -> Result<SpeechTranscriptionResult, String> {
    state
        .speech_service
        .finish_input_session(request)
        .await
        .map_err(|error| format!("Failed to transcribe speech input: {error}"))
}

#[tauri::command]
pub async fn speech_cancel_input_session(
    state: State<'_, AppState>,
    request: SpeechCancelInputSessionRequest,
) -> Result<(), String> {
    state
        .speech_service
        .cancel_input_session(request)
        .await
        .map_err(|error| format!("Failed to cancel speech input session: {error}"))
}

fn emit_status(app: &AppHandle, status: &SpeechModelStatus) {
    if let Err(error) = app.emit(SPEECH_MODEL_STATUS_CHANGED_EVENT, status) {
        log::warn!("Failed to emit speech model status event: {error}");
    }
}
